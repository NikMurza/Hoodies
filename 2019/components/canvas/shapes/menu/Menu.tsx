import * as React from 'react';
import { Group, Layer } from 'react-konva';
import MenuItem, { ICanvasMenuItem } from './MenuItem';

export interface ICanvasMenuProps {
  isVisible?: boolean;
  menu?: ICanvasMenuItem[];
  onClickItem?: (item: ICanvasMenuItem) => void;
  onLeave?: () => void;
  position?: number[];
}

interface IState {
  activeItems: number[];
}

export class Menu extends React.Component<ICanvasMenuProps, IState> {
  constructor(props) {
    super(props);
    this.state = {
      activeItems: [],
    };
  }

  /**
   * Рисуем уровни меню
   * @private
   */
  private getItems = () => {
    const { menu } = this.props;
    const { activeItems } = this.state;
    return menu.map((item, subIndex) => (
      <MenuItem
        index={subIndex}
        level={0}
        key={subIndex}
        item={item}
        activeItems={activeItems}
        onClick={this.handleActivateItem}
      />
    ));
  };

  private handleLeaveMenu = () => {
    this.setState({ activeItems: [] });
    this.props.onLeave();
  };

  private handleActivateItem = (level: number, index: number, item: ICanvasMenuItem) => {
    const { onClickItem } = this.props;
    const { activeItems } = this.state;
    const active = activeItems.slice(0, level);
    if (item.items) {
      active[level] = index;
    } else if (item.object) {
      onClickItem(item);
    }
    this.setState({ activeItems: active });
  };

  public render() {
    const { isVisible, position } = this.props;
    // если есть позиция то рисуем меню
    return (
      isVisible &&
      position && (
        <Layer>
          <Group x={position[0]} y={position[1]} onmouseleave={this.handleLeaveMenu}>
            {this.getItems()}
          </Group>
        </Layer>
      )
    );
  }
}

export default Menu;
