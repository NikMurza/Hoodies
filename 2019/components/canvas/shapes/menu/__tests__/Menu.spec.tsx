import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Menu from '../Menu';

describe('CanvasMenu component', () => {
  it('render default', async () => {
    expect(
      toJson(
        shallow(
          <Menu
            isVisible
            position={[100, 200]}
            menu={[
              {
                text: 'menu1',
                items: [
                  {
                    text: 'Menu2',
                  },
                ],
              },
            ]}
          />,
        ),
      ),
    ).toMatchSnapshot();
  });

  it('_handleLeaveMenu', async () => {
    const onLeaveHandler = jest.fn();
    const wrapper = shallow(
      <Menu
        isVisible
        position={[100, 200]}
        menu={[
          {
            text: 'menu1',
            items: [
              {
                text: 'Menu2',
              },
            ],
          },
        ]}
        onLeave={onLeaveHandler}
      />,
    );
    wrapper.setState({ activeItems: [0, 1] });
    wrapper.find('Group').get(0).props.onmouseleave();
    expect(wrapper.state('activeItems')).toEqual([]);
    expect(onLeaveHandler).toHaveBeenCalled();
  });

  it('_handleActivateItem case empty item', async () => {
    const wrapper = shallow(
      <Menu
        isVisible
        position={[100, 200]}
        menu={[
          {
            text: 'menu1',
            items: [
              {
                text: 'Menu2',
              },
            ],
          },
        ]}
      />,
    );
    wrapper.setState({ activeItems: [0, 1] });
    wrapper.find('MenuItem').get(0).props.onClick(1, 0, {});
    expect(wrapper.state('activeItems')).toEqual([0]);
  });

  it('_handleActivateItem case item has nested items', async () => {
    const wrapper = shallow(
      <Menu
        isVisible
        position={[100, 200]}
        menu={[
          {
            text: 'menu1',
            items: [
              {
                text: 'Menu2',
              },
            ],
          },
        ]}
      />,
    );
    wrapper.setState({ activeItems: [0] });
    wrapper
      .find('MenuItem')
      .get(0)
      .props.onClick(0, 5, { items: [{}, {}] });
    expect(wrapper.state('activeItems')).toEqual([5]);
  });

  it('_handleActivateItem case item has object', async () => {
    const onClickItem = jest.fn();
    const wrapper = shallow(
      <Menu
        isVisible
        position={[100, 200]}
        menu={[
          {
            text: 'menu1',
            items: [
              {
                text: 'Menu2',
              },
            ],
          },
        ]}
        onClickItem={onClickItem}
      />,
    );
    wrapper.setState({ activeItems: [0] });
    wrapper.find('MenuItem').get(0).props.onClick(0, 0, { object: {} });
    expect(onClickItem).toHaveBeenCalled();
    expect(wrapper.state('activeItems')).toEqual([]);
  });
});
