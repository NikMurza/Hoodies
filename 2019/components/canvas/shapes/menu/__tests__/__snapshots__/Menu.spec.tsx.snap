// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasMenu component render default 1`] = `
<Layer>
  <Group
    onmouseleave={[Function]}
    x={100}
    y={200}
  >
    <MenuItem
      activeItems={Array []}
      index={0}
      item={
        Object {
          "items": Array [
            Object {
              "text": "Menu2",
            },
          ],
          "text": "menu1",
        }
      }
      key="0"
      level={0}
      offsetX={0}
      offsetY={0}
      onClick={[Function]}
    />
  </Group>
</Layer>
`;
